### DISCUSSION TITLE
Creating a minecart computer.

### DISCUSSION THREAD

**Discussion by vexxmage** [cc-tweaked/CC-Tweaked#1221] [Q&A] [answered]
Hi, I am asking for a bit of direction on where to look on in the code that would help me create a minecart that is also a computer.

I am new to creating a mod, but I have already got a minecart where irs speed can be changed in gui.

Any help would be appreciated.

**Comment by SquidDev** [comment, marked as answer]
I should probably start off with a "here be dragons". The public API doesn't support creating new types of computer, and so anything you do will involve poking the mods internals (which, of course, doesn't come with any stability guarantees).

The best thing to do is probably start by reading through the computer `BlockEntity` code (https://github.com/cc-tweaked/CC-Tweaked/blob/8d2e150f05b49eb96fb37d40746accad503db077/projects/common/src/main/java/dan200/computercraft/shared/computer/blocks/AbstractComputerBlockEntity.java). The general gist is:
 - Obtain a `ServerComputer` (see `createServerComputer`): This is the object responsible for managing the underlying computer, as well as syncing its state to the client, etc...
 - On each tick, call `ServerComputer.keepAlive();` and sync its state back to your entity. You'll want to make sure that the computer ID, label and on/off state are persisted.
 - When the computer is right clicked, open the UI using `new ComputerContainerData(computer, stack).open(player, this);`

**Reply by vexxmage** [reply]
Thank you, that was very helpful. 
I did suspect that I couldn't do it with the API.
I am a software engineer, so I am unfortunately use to dealing with code that is changing a lot. 

I was able to create the minecart computer and create a minecart API to sent it commands.
It was a lot easier than I was expecting.
